File: javascript/packages/config/test/find-project-root.test.ts
import { describe, test, expect, beforeAll, afterAll } from "vitest"
import { mkdirSync, writeFileSync, rmSync } from "fs"
import { join } from "path"
import { Config } from "../src/config.js"

describe("findProjectRootSync with .herb.yaml", () => {
  const tempDir = join(process.cwd(), "tmp-test-find-project-root-yaml")

  beforeAll(() => {
    mkdirSync(join(tempDir, "app", "views"), { recursive: true })

    writeFileSync(join(tempDir, ".herb.yaml"), "linter:\n  enabled: true\n")
    writeFileSync(join(tempDir, "app", "views", "index.html.erb"), "<div></div>\n")
  })

  afterAll(() => {
    rmSync(tempDir, { recursive: true, force: true })
  })

  test("finds project root with .herb.yaml", () => {
    const filePath = join(tempDir, "app", "views", "index.html.erb")
    const projectRoot = Config.findProjectRootSync(filePath)

    expect(projectRoot).toBe(tempDir)
  })
})

describe("findProjectRootSync", () => {
  const tempDir = join(process.cwd(), "tmp-test-find-project-root")

  beforeAll(() => {
    mkdirSync(join(tempDir, "app", "frontend", "components"), { recursive: true })
    mkdirSync(join(tempDir, "app", "views", "layouts"), { recursive: true })

    writeFileSync(join(tempDir, ".herb.yml"), "linter:\n  exclude:\n    - 'app/views/**/*.html.erb'\n")
    writeFileSync(join(tempDir, "app", "frontend", "README.md"), "# Frontend\n")
    writeFileSync(join(tempDir, "app", "frontend", "components", "button.html.erb"), "<button></button>\n")
    writeFileSync(join(tempDir, "app", "views", "layouts", "application.html.erb"), "<html></html>\n")
  })

  afterAll(() => {
    rmSync(tempDir, { recursive: true, force: true })
  })

  test("finds project root with .herb.yml over subdirectory with README.md", () => {
    const filePath = join(tempDir, "app", "frontend", "components", "button.html.erb")
    const projectRoot = Config.findProjectRootSync(filePath)

    expect(projectRoot).toBe(tempDir)
  })

  test("finds project root from deeply nested file path", () => {
    const filePath = join(tempDir, "app", "views", "layouts", "application.html.erb")
    const projectRoot = Config.findProjectRootSync(filePath)

    expect(projectRoot).toBe(tempDir)
  })

  test("prefers .herb.yml over soft project indicators like README.md", () => {
    const filePath = join(tempDir, "app", "frontend", "components", "button.html.erb")
    const projectRoot = Config.findProjectRootSync(filePath)

    expect(projectRoot).not.toBe(join(tempDir, "app", "frontend"))
    expect(projectRoot).toBe(tempDir)
  })
})

File: rust/herb-config/tests/find_project_root_test.rs
use std::fs;
use std::path::PathBuf;

use herb_config::Config;

struct Fixture {
  _dir: tempfile::TempDir,
  root: PathBuf,
}

impl Fixture {
  fn new() -> Self {
    let dir = tempfile::tempdir().unwrap();
    let root = fs::canonicalize(dir.path()).unwrap();

    fs::create_dir_all(root.join("app/frontend/components")).unwrap();
    fs::create_dir_all(root.join("app/views/layouts")).unwrap();

    fs::write(root.join(".herb.yml"), "linter:\n  exclude:\n    - 'app/views/**/*.html.erb'\n").unwrap();
    fs::write(root.join("app/frontend/README.md"), "# Frontend\n").unwrap();
    fs::write(root.join("app/frontend/components/button.html.erb"), "<button></button>\n").unwrap();
    fs::write(root.join("app/views/layouts/application.html.erb"), "<html></html>\n").unwrap();

    Self { _dir: dir, root }
  }
}

#[test]
fn finds_project_root_with_a_herb_yml_over_subdirectory_with_readme_md() {
  let fixture = Fixture::new();
  let file_path = fixture.root.join("app/frontend/components/button.html.erb");

  assert_eq!(Config::find_project_root_sync(&file_path), fixture.root);
}

#[test]
fn finds_project_root_from_deeply_nested_file_path() {
  let fixture = Fixture::new();
  let file_path = fixture.root.join("app/views/layouts/application.html.erb");

  assert_eq!(Config::find_project_root_sync(&file_path), fixture.root);
}

#[test]
fn prefers_a_herb_yml_over_soft_project_indicators_like_readme_md() {
  let fixture = Fixture::new();
  let file_path = fixture.root.join("app/frontend/components/button.html.erb");

  let project_root = Config::find_project_root_sync(&file_path);

  assert_ne!(project_root, fixture.root.join("app/frontend"));
  assert_eq!(project_root, fixture.root);
}
